File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.cpp
class Solution {
private:
    int binSearch(int left, int right, int target, const vector<vector<int>>& events) {
        if (left >= right) {
            return right;
        }
        
        int mid = left + (right - left) / 2;
        if (events[mid][0] > target) {
            return binSearch(left, mid, target, events);
        } else {
            return binSearch(mid + 1, right, target, events);
        }
    }
    
public:
    int maxTwoEvents(vector<vector<int>>& events) {
        sort(events.begin(), events.end(), 
             [](const vector<int>& a, const vector<int>& b) {
                 return a[0] < b[0];
             });
        
        int n = events.size();
        vector<int> maxSuffix(n);
        
        maxSuffix[n - 1] = events[n - 1][2];
        for (int i = n - 2; i >= 0; i--) {
            maxSuffix[i] = max(maxSuffix[i + 1], events[i][2]);
        }
        
        int maxValue = 0;
        for (int i = 0; i < n; i++) {
            int currentValue = events[i][2];
            int nextEventIndex = binSearch(i + 1, n, events[i][1], events);
            int additionalValue = (nextEventIndex < n) ? maxSuffix[nextEventIndex] : 0;
            maxValue = max(maxValue, currentValue + additionalValue);
        }
        
        return maxValue;
    }
};

File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.java
import java.util.*;

class Solution {
    public int maxTwoEvents(int[][] events) {
        Arrays.sort(events, (a, b) -> a[1] - b[1]);

        int n = events.length;
        int[] endT = new int[n];
        int[] best = new int[n];

        for (int i = 0; i < n; i++) {
            endT[i] = events[i][1];
            best[i] = events[i][2];
            if (i > 0) best[i] = Math.max(best[i], best[i - 1]);
        }

        int ans = 0;

        for (int i = 0; i < n; i++) {
            int st = events[i][0];
            int val = events[i][2];

            int l = 0, r = n - 1, idx = -1;
            while (l <= r) {
                int mid = (l + r) >>> 1;
                if (endT[mid] < st) {
                    idx = mid;
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }

            if (idx != -1) ans = Math.max(ans, val + best[idx]);
            ans = Math.max(ans, val);
        }

        return ans;
    }
}